cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -300,30 +395,36 @@ class HashJoinNode : public ExecNode {
 
     const auto& join_options = checked_cast<const HashJoinNodeOptions&>(options);
 
+    const auto& left_schema = *(inputs[0]->output_schema());
+    const auto& right_schema = *(inputs[1]->output_schema());
     // This will also validate input schemas
     if (join_options.output_all) {
       RETURN_NOT_OK(schema_mgr->Init(
-          join_options.join_type, *(inputs[0]->output_schema()), join_options.left_keys,
-          *(inputs[1]->output_schema()), join_options.right_keys,
+          join_options.join_type, left_schema, join_options.left_keys, right_schema,
+          join_options.right_keys, join_options.filter,
           join_options.output_prefix_for_left, join_options.output_prefix_for_right));
     } else {
       RETURN_NOT_OK(schema_mgr->Init(
-          join_options.join_type, *(inputs[0]->output_schema()), join_options.left_keys,
-          join_options.left_output, *(inputs[1]->output_schema()),
-          join_options.right_keys, join_options.right_output,
+          join_options.join_type, left_schema, join_options.left_keys,
+          join_options.left_output, right_schema, join_options.right_keys,
+          join_options.right_output, join_options.filter,
           join_options.output_prefix_for_left, join_options.output_prefix_for_right));
     }
 
+    ARROW_ASSIGN_OR_RAISE(Expression filter,
+                          schema_mgr->BindFilter(std::move(join_options.filter),
+                                                 left_schema, right_schema));
+
```

Comment:
I think you'd get away with it 9 times out of 10 but I'm not sure it is safe to move `join_options.filter`